OpenRA.Mods.Common/Traits/World/Locomotor.cs

```diff
@@ -445,32 +455,52 @@ void UpdateCellBlocking(CPos cell)
 
 				var cellBits = default(LongBitSet<PlayerBitMask>);
 
+				var cellIsMoving = true;
+				var cellHasTemporaryBlocker = true;
+				var cellIsCrushable = true;
+
 				foreach (var actor in actors)
 				{
-					var actorBits = default(LongBitSet<PlayerBitMask>);
+					var actorBits = world.AllPlayerMask;
 					var crushables = actor.TraitsImplementing<ICrushable>();
 					var mobile = actor.OccupiesSpace as Mobile;
-					var isMoving = mobile != null && mobile.CurrentMovementTypes.HasFlag(MovementType.Horizontal);
+					var isMoving = mobile != null && mobile.CurrentMovementTypes.HasMovementType(MovementType.Horizontal);
 
 					if (crushables.Any())
 					{
-						LongBitSet<PlayerBitMask> crushingBits;
 						foreach (var crushable in crushables)
-							if (crushable.TryCalculatePlayerBlocking(actor, Info.Crushes, out crushingBits))
-							{
-								actorBits = actorBits.Union(crushingBits);
-								cellBlocking |= CellBlocking.Crushable;
-							}
-
-						if (isMoving)
-							actorBits = actorBits.Except(actor.Owner.AllyMask);
+						{
+                            actorBits = actorBits.Union(crushable.CrushableBy(actor, Info.Crushes));
+						}
 					}
 					else
-						actorBits = isMoving ? actor.Owner.EnemyMask : allPlayerMask;
+						cellIsCrushable = false;
+
+					if (!isMoving)
+						cellIsMoving = false;
+
+					// PERF: Only perform ITemporaryBlocker trait look-up if mod/map rules contain any actors that are temporary blockers
+					if (world.RulesContainTemporaryBlocker)
+					{
+						// If there is a temporary blocker in this cell.
+						if (actor.TraitOrDefault<ITemporaryBlocker>() == null)
+							cellHasTemporaryBlocker = false;
```

Comment:
Shouldn't this be false only if *all* actors in the cell don't have the interface?